Add ability to add/remove/review Account Admins in SuperAdmin UI

tdonohue opened the pull request:
Fixes #725 by enhancing the existing SuperAdmin UI (`/proprietor/accounts`) to provide the ability to easily add/remove/review Account Administrators.  Here's an example screenshot (for an account with cname="test3.lvh.me"):
![screenshot](https://user-images.githubusercontent.com/483997/26890630-da40e5ee-4b77-11e7-905e-105e84d3cbf6.png)

Adding a new Account Administrator uses `devise_invitable` to send an email invitation if the person does not yet have a login for that Account. If the person does have a login, then their login is immediately updated with Admin rights.

Also cleans up existing SuperAdmin UI screens to use `simple_form_for` and add i18n to all views.

**NOTE: This is flagged as a WIP, as it does not yet include updated specs or translations.**  However, the basic functionality is ready for review/comment, as it works.

@projecthydra-labs/hyrax-code-reviewers

ggeisler commented:
@tdonohue Just commenting based on the screenshot: looks good! I have a couple of very minor suggestions for updating the table column headings:

- Email > Email address
- Actions > Action (assuming that Remove is the only action we expect to show in that column)

Also, I usually avoid using the word "Please" in the UI; it doesn't really add anything useful and omitting it helps keep the UI simpler and more direct.

mjgiarlo commented on app/models/account.rb:
```diff
@@ -95,6 +96,28 @@ def reset!
+    # Must ensure we are switched to proper tenant database
+    Apartment::Tenant.switch(tenant) do
+      # return list of email addresses of all users with admin role on this site
+      User.with_role(:admin, Site.instance).map(&:email)
```
There may be some small amount of optimization available here by using `.pluck(:email)` instead of `.map(&:email)`, since that moves the operation into the DB query.

mjgiarlo replied:
I wonder if we may want to relocate this logic somewhere else, like in the `Site` model.

mjgiarlo commented on config/locales/devise_invitable.en.yml:
```diff
@@ -0,0 +1,31 @@
+en:
```
Should we add Spanish and Chinese translations too?

tdonohue replied:
Done

mjgiarlo commented on config/locales/en.yml:
```diff
@@ -121,3 +121,28 @@ en:
         roles_and_permissions:        'Users and groups'
         system_status:                'System status'
         technical:                    'Technical'
+  proprietor:
```
Can you run this on Hyku once all the translations have been added: https://github.com/samvera-labs/hyrax/commit/dfe0c350bb1f5db9a6c01e186ecb852b5ab8c994#diff-54441a884831d33711a0af9d92a20a32

This lets us keep our locale files looking similar.

tdonohue replied:
@mjgiarlo : I have not done this, as I'm having difficulty figuring out a way to run this Hyrax script against Hyku files...it doesn't seem to be included in Hyku (unless I'm missing it).

mjgiarlo replied:
No, you're right, @tdonohue, and I now realize that we do *not* sort our YAML files in Hyku. Maybe something for a later date. Thanks!

mjgiarlo commented on db/migrate/20170605152346_devise_invitable_add_to_users.rb:
```diff
@@ -0,0 +1,12 @@
+    add_column :users, :invitation_limit, :integer
+    add_column :users, :invited_by_id, :integer
+    add_column :users, :invited_by_type, :string
+    add_index :users, :invitation_token, :unique => true
```
I wonder why Rubocop didn't balk at the :hash: :rocket: here.

jcoyne commented:
This looks great. 👍 to the translation suggestions that Mike made.

tdonohue commented:
Taking WIP off of this as I've now completed the following:
* Added specs
* Added translations (Spanish & Chinese) for all new i18n keys
* Tweaked UI per @ggeisler's suggestions
* Refactored/updated per @mjgiarlo's suggestions, *except* for the reordering (alphabetizing) of i18n keys, as that Hyrax tool seems to not be available in Hyku

So, I think this is ready for review again.

mjgiarlo commented on app/models/account.rb:
```diff
@@ -95,6 +95,22 @@ def reset!
   end
 
+  # Get admin emails associated with this account/site
+  def admin_emails
```
Can you add coverage for the changes to this file to `spec/models/account_spec.rb`?

tdonohue replied:
Good catch. Can do

tdonohue replied:
Done now

tdonohue commented:
Hi @mjgiarlo : Just as a note, the changes to `new`, `edit` and `index` were strictly refactoring to use `simple_form_for` and `i18n`, which is why I didn't update those existing specs. Essentially, those pages did not change in structure or content, but were refactored to be more like others in Hyku.  You can see the entirety of those changes in this commit: https://github.com/samvera-labs/hyku/pull/1218/commits/43bf5d44252259e1353bd68fa9ae5cd2fbfafef9

darrenleeweber commented on app/models/account.rb:
```diff
@@ -95,6 +95,22 @@ def reset!
+    Apartment::Tenant.switch(tenant) do
+      Site.instance.admin_emails = emails
+    end
+  end
```
Looking at the allowed params to `new`, it seems that the `admin_emails` is an array of emails, but this is not entirely clear that an array of emails is the input to this method when setting the values here or returning them in the getter.  (That is, there is no documentation on the method, as in yard docs with @param attributes.)  So, I'm left wondering whether we are working with arrays of strings or maybe CSV strings or what?

darrenleeweber commented on app/models/site.rb:
```diff
@@ -18,4 +18,40 @@ def instance
+    removed_admin_emails = existing_admin_emails - emails
+    add_admins_by_email(new_admin_emails) if new_admin_emails
+    remove_admins_by_email(removed_admin_emails) if removed_admin_emails
+  end
```
This method is clearly documented to clarify that we're working with an `Array<String>` 👍

darrenleeweber commented on app/models/site.rb:
```diff
@@ -18,4 +18,40 @@ def instance
+  end
+
+  private
+
```
This method (below) appears to accept only one email, as in `@param [String] email` but the name of the param implies a plural array of emails.  Could the param name be singular and add an `@param` docstring to clarify the intended param?

mjgiarlo commented:
Thanks for the clarification, @tdonohue!

darrenleeweber commented on spec/models/site_spec.rb:
```diff
@@ -1,7 +1,55 @@
+      end
+      it "returns array of emails" do
+        expect(subject.admin_emails).to match_array([admin1.email, admin2.email])
+      end
```
Is `match_array` sensitive to the order of the elements in the array?  Is the order important?

darrenleeweber replied:
I see that `match_array` is not senstive to the order of elements in the array (I would argue that they should have called this matcher `match_set` instead), see
- http://www.rubydoc.info/github/rspec/rspec-expectations/RSpec/Matchers#contain_exactly-instance_method

tdonohue replied:
Thanks for the comments, @darrenleeweber. From my understanding `match_array` is not order dependent, e.g. https://stackoverflow.com/a/11796957/3750035

tdonohue commented:
Per @mjgiarlo's and @darrenleeweber's comments above, I've added a few missing Specs (for `account`) and enhanced the comments around method parameters. So, this is ready for another review

mjgiarlo commented:
I changed my mind -- this is good to go, IMO.
